### PR TITLE
Certificate Revocation Validation with CRL and OCSP

### PR DESCRIPTION
**Pull request by IndunilRathnayake**
## Purpose
X509 Certificate Revocation Validation with CRL and OCSP

### CODE REVIEW

File: component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/service/RevocationValidationManagerImpl.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        Collections.sort(revocationValidators, revocationValidatorComparator);
+        int validatorCount = revocationValidators.size();
+
+        for (RevocationValidator validator : revocationValidators) {
+            --validatorCount;
+            if (validator.isEnable()) {
+                try {
+                    return checkValidity(validator, peerCertificate);
+                } catch (CertificateValidationException e) {
+                    if (validatorCount <= 0) {
```
Comment (IndunilRathnayake) [inline component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/service/RevocationValidationManagerImpl.java]: mention the validator and cert in a log

File: component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/service/RevocationValidationManagerImpl.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                }
+                return checkValidity(validator, caCertificate.getX509Certificate());
+            } else if (RevocationStatus.GOOD.equals(revocationStatus)) {
+                if(log.isDebugEnabled()) {
+                    log.debug("Certificate with serial num: " + certificate.getSerialNumber() + " is not revoked.");
+                }
+                return false;
+            }
+        }
+        throw new CertificateValidationException("Cannot check revocation status of the certificate.");
```
Comment (IndunilRathnayake) [inline component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/service/RevocationValidationManagerImpl.java]: validator and certificate info to the exception

File: component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java [hunk 3]
```diff
@@ -0,0 +1,1054 @@
...
+        byte[] crlDPExtensionValue = getCRLDPExtensionValue(cert);
+        if (crlDPExtensionValue == null) {
+            log.error("Certificate doesn't have CRL Distribution points");
+            return crlUrls;
+        }
+        CRLDistPoint distPoint = getCrlDistPoint(crlDPExtensionValue);
+        getCrlUrlsFromDistPoint(crlUrls, distPoint);
+
+        if (crlUrls.isEmpty()) {
+            throw new CertificateValidationException("Cant get CRL urls from certificate");
```
Comment (IndunilRathnayake) [inline component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java]: don't throw the exception

File: component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java [hunk 4]
```diff
@@ -0,0 +1,1054 @@
...
+     * @return List of CRL Urls in the certificate
+     * @throws CertificateValidationException certificateValidationException
+     */
+    public static List<String> getCRLUrls(X509Certificate cert)
+            throws CertificateValidationException {
+
+        List<String> crlUrls = new ArrayList<>();
+        byte[] crlDPExtensionValue = getCRLDPExtensionValue(cert);
+        if (crlDPExtensionValue == null) {
+            log.error("Certificate doesn't have CRL Distribution points");
```
[the remark this answers is not in the document]
Comment (IndunilRathnayake) [reply component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java]: throw an exception

File: component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java [hunk 5]
```diff
@@ -0,0 +1,1054 @@
...
+            if (retryCount == 0) {
+                throw new CertificateValidationException("Cant reach URI: " + crlURL + " - only support HTTP", e);
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Cant reach URI: " + crlURL + ". Retrying to connect - attempt " + retryCount);
+                }
+                downloadCRLFromWeb(crlURL, --retryCount);
+            }
+        } catch (CertificateException e) {
+            throw new CertificateValidationException(e);
```
Comment (IndunilRathnayake) [inline component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java]: exception message

File: component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java [hunk 6]
```diff
@@ -0,0 +1,1054 @@
...
+            x509CRL = (X509CRL) cf.generateCRL(crlStream);
+
+            if (log.isDebugEnabled()) {
+                log.debug("CRL is downloaded from CRL Url: " + crlURL);
+            }
+        } catch (MalformedURLException e) {
+            throw new CertificateValidationException("CRL Url is malformed", e);
+        } catch (IOException e) {
+            if (retryCount == 0) {
+                throw new CertificateValidationException("Cant reach URI: " + crlURL + " - only support HTTP", e);
```
Comment (IndunilRathnayake) [inline component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java]: check http only

File: component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java [hunk 7]
```diff
@@ -0,0 +1,1054 @@
...
+                        return status;
+                    } else {
+                        if (log.isDebugEnabled()) {
+                            log.debug("CRL is too old. Removing from cache.");
+                        }
+                        CRLCache.getInstance().clearCacheEntry(crlUrl);
+                    }
+                }
+
+                x509CRL = downloadCRLFromWeb(crlUrl, retryCount);
```
Comment (IndunilRathnayake) [inline component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java]: next update field validation

File: component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java [hunk 8]
```diff
@@ -0,0 +1,1054 @@
...
+        for (String crlUrl : crlUrls) {
+            if (log.isDebugEnabled()) {
+                log.debug("Trying to get CRL for URL: " + crlUrl);
+            }
+
+            X509CRL x509CRL = getCRLFromCache(crlUrl);
+            try {
+                if (x509CRL != null) {
+                    if (isValidCrlFromCache(x509CRL)) {
+                        RevocationStatus status = getRevocationStatusFromCRL(x509CRL, peerCert);
```
Comment (IndunilRathnayake) [inline component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java]: just return the status

File: component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java [hunk 9]
```diff
@@ -0,0 +1,1054 @@
...
+        byte[] aiaExtensionValue = getAiaExtensionValue(cert);
+        if (aiaExtensionValue == null) {
+            log.error("Certificate Doesn't have Authority Information Access points");
+            return ocspUrlList;
+        }
+        AuthorityInformationAccess authorityInformationAccess = getAuthorityInformationAccess(aiaExtensionValue);
+        getOcspUrlsFromAuthorityInfoAccess(ocspUrlList, authorityInformationAccess);
+
+        if (ocspUrlList.isEmpty())
+            throw new CertificateValidationException("Cant get OCSP urls from certificate");
```
Comment (IndunilRathnayake) [inline component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java]: put a bracket

File: component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java [hunk 10]
```diff
@@ -0,0 +1,1054 @@
...
+     * @return Revocation status of the certificate
+     * @throws CertificateValidationException certificateValidationException
+     */
+    public static RevocationStatus getRevocationStatus(X509Certificate peerCert, X509Certificate issuerCert,
+                                                       int retryCount, List<String> locations)
+            throws CertificateValidationException {
+
+        OCSPReq request = generateOCSPRequest(issuerCert, peerCert.getSerialNumber());
+        if (CollectionUtils.isNotEmpty(locations)) {
+            for (String serviceUrl : locations) {
```
Comment (IndunilRathnayake) [inline component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java]: isNotEmpty is not needed.

File: component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java [hunk 11]
```diff
@@ -0,0 +1,1054 @@
...
+                    }
+
+                    BasicOCSPResp basicResponse = (BasicOCSPResp) ocspResponse.getResponseObject();
+                    responses = (basicResponse == null) ? null : basicResponse.getResponses();
+                } catch (Exception e) {
+                    continue;
+                }
+
+                if (responses != null && responses.length == 1) {
+                    SingleResp resp = responses[0];
```
Comment (IndunilRathnayake) [inline component/validation/src/main/java/org/wso2/carbon/identity/x509Certificate/validation/CertificateValidationUtil.java]: put responses[0] direclty

**Review by isharak — APPROVED** [review]

**Review by isharak — APPROVED** [review]

**Pull request merged**